File: 18. 4Sum/Solution.py
class Solution:
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        ans = []
        if len(nums) < 4:
            return ans
        nums.sort()
        for i in range(len(nums) - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            if sum(nums[i:i + 4]) > target:
                break
            if nums[i] + sum(nums[-3:]) < target:
                continue
            for j in range(i + 1, len(nums) - 2):
                k = j + 1
                l = len(nums) - 1
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                if nums[i] + sum(nums[j:j + 3]) > target:
                    break
                if nums[i] + nums[j] + sum(nums[-2:]) < target:
                    continue
                while k < l:
                    sum1 = nums[i] + nums[j] + nums[k] + nums[l]
                    if sum1 == target:
                        ans.append([nums[i], nums[j], nums[k], nums[l]])
                        k += 1
                        l -= 1
                        while k < l and nums[k] == nums[k - 1]:
                            k += 1
                        while k < l and nums[l] == nums[l + 1]:
                            l -= 1
                    elif sum1 < target:
                        k += 1
                        while k < l and nums[k] == nums[k - 1]:
                            k += 1
                    else:
                        l -= 1
                        while k < l and nums[l] == nums[l + 1]:
                            l -= 1
        return ans

File: 18. 4Sum/Solution.cpp
class Solution
{
  public:
    vector<vector<int>> fourSum(vector<int> &nums, int target)
    {
        vector<vector<int>> ans;
        int len = nums.size();
        if (len < 4)
            return ans;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < len - 3; i++)
        {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            if (nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target)
                break;
            if (nums[i] + nums[len - 3] + nums[len - 2] + nums[len - 1] < target)
                continue;
            for (int j = i + 1; j < len - 2; j++)
            {
                int k = j + 1, l = len - 1;
                if (j > i + 1 && nums[j] == nums[j - 1])
                    continue;
                if (nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target)
                    break;
                if (nums[i] + nums[j] + nums[len - 2] + nums[len - 1] < target)
                    continue;
                while (k < l)
                {
                    int sum = nums[i] + nums[j] + nums[k] + nums[l];
                    if (sum == target)
                    {
                        ans.push_back({nums[i], nums[j], nums[k], nums[l]});
                        k++;
                        while (k < l && nums[k] == nums[k - 1])
                            k++;
                        l--;
                        while (k < l && nums[l] == nums[l + 1])
                            l--;
                    }
                    else if (sum < target)
                    {
                        k++;
                        while (k < l && nums[k] == nums[k - 1])
                            k++;
                    }
                    else
                    {
                        l--;
                        while (k < l && nums[l] == nums[l + 1])
                            l--;
                    }
                }
            }
        }
        return ans;
    }
};